GAE/src/org/akvo/flow/servlet/ReportServlet.java

```diff
@@ -0,0 +1,237 @@
+
+                        if (sts == 200) {
+                            //Success, we are done!
+                            return null;
+                        } else if ((sts % 100) == 4) { //4xx: you messed up
+                            //permanent error, fail this report
+                            r.setState(Report.FINISHED_ERROR);
+                            r.setMessage("Unexpected result when starting report: " + sts);
+                            rDao.save(r);
+                        } else {
+                            //if we get a transient error, re-queue
+                            requeueStart(r);
+                        }
+                    }
+                    catch (MalformedURLException e) {
+                        Log.error("Bad URL");
+                    }
+                    catch (IOException e) {
+                        log.warning("====IOerror: " + e);
+                        //call it a transient error, re-queue
+                        requeueStart(r);
+                    }
+                }
+                break;
+            case ReportTaskRequest.PROGRESS_ACTION:
+                Report r2 = rDao.getByKey(id);
+                if (r2 != null) {
+                    if (!r2.getState().equals(Report.QUEUED)
+                            && !r2.getState().equals(Report.IN_PROGRESS)) {
+                        //wrong state
+                        log.warning("Cannot set progress on report that is " + r2.getState());
+                        return null;
+                    }
+                    r2.setState(stReq.getState());
+                    r2.setMessage(stReq.getMessage());
+                    rDao.save(r2);
+                }
+                break;
+            default:
+                log.warning("Unknown action.");
+                break;
+        }
+        return null;
+    }
+
+    public static void queueStart(Report r) {
+        log.info("Forking to task with action START");
+        Queue queue = QueueFactory.getDefaultQueue();
+        TaskOptions options = TaskOptions.Builder.withUrl(SERVLET_URL)
+                .param(TaskRequest.ACTION_PARAM, ReportTaskRequest.START_ACTION)
+                .param(ReportTaskRequest.ID_PARAM, Long.toString(r.getKey().getId()));
+        queue.add(options);
+    }
+
+    private void requeueStart(Report r) {
+        log.warning("Requeuing task with action START");
+        //TODO give up if this has been going on too long
+
+        Queue queue = QueueFactory.getDefaultQueue();
+        TaskOptions options = TaskOptions.Builder.withUrl(SERVLET_URL)
+                .param(TaskRequest.ACTION_PARAM, ReportTaskRequest.START_ACTION)
+                .param(ReportTaskRequest.ID_PARAM, Long.toString(r.getKey().getId()))
+                .countdownMillis(Constants.TASK_RETRY_INTERVAL);
+        queue.add(options);
+    }
+
+    private int engage(Report r) throws JsonGenerationException, JsonMappingException, IOException {
```

Comment:
I was trying to figure out what `engage()` was. 😄 maybe a more descriptive name?